File: sdk/node/_fingerprint.js
// Public switch syntax; keep in parity with the Python launcher.
const OFF = new Set(["off", "false", "0", "disable", "disabled"]);
const BRANDS = new Map([["chrome", "Chrome"], ["google chrome", "Chrome"],
  ["edge", "Edge"], ["microsoft edge", "Edge"], ["opera", "Opera"], ["vivaldi", "Vivaldi"]]);
const BOOLEANS = new Set(["--fingerprint-noise", "--fingerprint-sapi-voices",
  "--fingerprint-allow-3p-cookies", "--fingerprint-windows-font-metrics"]);
const INTEGERS = new Map([
  ["--fingerprint-hardware-concurrency", [1n, 128n]],
  ["--fingerprint-screen-width", [1n, 32768n]],
  ["--fingerprint-screen-height", [1n, 32768n]],
  ["--fingerprint-taskbar-height", [0n, 32768n]],
  ["--fingerprint-storage-quota", [0n, ((1n << 63n) - 1n) / 1048576n]],
]);
const PREFERENCE_BOOLEANS = new Set(["reduced-motion", "reduced-transparency", "inverted-colors"]);
const PREFERENCE_INTEGERS = new Map([["max-touch-points", [0n, 16n]], ["timer-resolution", [0n, 1000n]],
  ["audio-seed", [1n, (1n << 64n) - 1n]]]);
const PREFERENCE_ENUMS = new Map(Object.entries({
  "gpu-backend": ["native", "compatibility"],
  "font-policy": ["native", "restricted"],
  "audio-render": ["native", "isolated"],
  "color-scheme": ["dark", "light"],
  "preferred-contrast": ["more", "less", "no-preference"],
  "forced-colors": ["active", "none", "true", "false", "1", "0"],
  pointer: ["fine", "coarse", "none"], hover: ["hover", "none"], hdr: ["native"],
  "keyboard-layout": ["native", "us", "en-US"],
}));
const CODECS = new Set(["codec-h264", "codec-vp8", "codec-vp9", "codec-av1", "codec-hevc"]);

function validCodec(value) {
  if (["", "native", "disabled"].includes(value)) return true;
  const tokens = value.split(",");
  return tokens.includes("supported") && tokens.every((t) => ["supported", "smooth", "power-efficient"].includes(t));
}

function validatePreferences(args, final) {
  const switches = new Map(args.map((arg) => {
    const i = arg.indexOf("=");
    return i < 0 ? [arg, ""] : [arg.slice(0, i), arg.slice(i + 1)];
  }));
  const value = (name) => switches.get(`--uxr-${name}`) ?? switches.get(`--fingerprint-${name}`);
  const pointer = value("pointer"), hover = value("hover");
  const touch = value("max-touch-points") === undefined ? undefined : Number(value("max-touch-points"));
  if ((pointer === "none" && touch > 0) || (pointer === "coarse" && touch === 0) ||
      (pointer === "none" && hover === "hover"))
    throw new Error("inconsistent pointer, hover and max-touch-points");
  const keyboard = value("keyboard-layout");
  if (keyboard !== undefined && keyboard !== "native" && switches.get("--uxr-synthetic-device-tests") !== "true")
    throw new Error("keyboard-layout overrides require synthetic device tests; use native for real input");
  if (value("font-policy") === "restricted") {
    const families = value("font-whitelist") || "", entries = families.split(",");
    if (!families || Buffer.byteLength(families, "utf8") > 4096 || entries.length > 256 ||
        entries.some((entry) => !entry.replace(/^[ \t]+|[ \t]+$/g, "")) || /[\x00-\x08\x0a-\x1f\x7f]/.test(families))
      throw new Error("restricted font-policy requires 1 to 256 comma-separated installed font families");
  }
  if (final && value("audio-render") === "isolated" && !fingerprintOff(args)) {
    const seed = value("audio-seed") ?? switches.get("--uxr-fingerprint-seed");
    if (seed === undefined) {
      if (!switches.has("--fingerprint"))
        throw new Error("isolated audio requires --fingerprint or a nonzero audio seed");
    } else if (!/^[0-9]+$/.test(seed) || BigInt(seed) < 1n || BigInt(seed) >= (1n << 64n)) {
      throw new Error("isolated audio requires a nonzero uint64 seed");
    }
  }
}

export function fingerprintOff(args) {
  const flag = args.filter((a) => a.split("=", 1)[0] === "--fingerprint").at(-1);
  return flag !== undefined && OFF.has(flag.slice(flag.indexOf("=") + 1).toLowerCase());
}

export function normalizeFingerprintArgs(args = [], { final = false } = {}) {
  let result = (args || []).map((arg) => {
    if (typeof arg !== "string" || arg.includes("\0"))
      throw new Error("browser arguments must be strings without NUL");
    const separator = arg.indexOf("=");
    const key = separator < 0 ? arg : arg.slice(0, separator);
    const value = separator < 0 ? "" : arg.slice(separator + 1);
    const preference = key.startsWith("--fingerprint-") ? key.slice(14) : key.startsWith("--uxr-") ? key.slice(6) : "";
    if (PREFERENCE_BOOLEANS.has(preference)) {
      const low = value.toLowerCase();
      if (OFF.has(low)) return `${key}=false`;
      if (["", "true", "1", "on", "enable", "enabled"].includes(low)) return `${key}=true`;
      throw new Error(`${key} requires a boolean`);
    } else if (PREFERENCE_INTEGERS.has(preference)) {
      const [min, max] = PREFERENCE_INTEGERS.get(preference);
      if (!/^[0-9]+$/.test(value) || BigInt(value) < min || BigInt(value) > max)
        throw new Error(`${key} requires an integer in [${min}, ${max}]`);
    } else if (PREFERENCE_ENUMS.has(preference)) {
      if (!PREFERENCE_ENUMS.get(preference).includes(value))
        throw new Error(`${key} requires one of ${PREFERENCE_ENUMS.get(preference).join(", ")}`);
    } else if (CODECS.has(preference)) {
      if (!validCodec(value))
        throw new Error(`${key} requires native, disabled or supported[,smooth][,power-efficient]`);
    } else if (key === "--fingerprint") {
      if (OFF.has(value.toLowerCase())) return "--fingerprint=off";
      if (value && (!/^[0-9]+$/.test(value) || BigInt(value) < 1n || BigInt(value) >= 1n << 64n))
        throw new Error("--fingerprint requires a uint64 seed or off/false/0/disable/disabled");
    } else if (key === "--fingerprint-brand") {
      const brand = BRANDS.get(value.toLowerCase());
      if (!brand) throw new Error("--fingerprint-brand must be Chrome, Edge, Opera or Vivaldi");
      return `${key}=${brand}`;
    } else if (BOOLEANS.has(key)) {
      const low = value.toLowerCase();
      if (OFF.has(low)) return `${key}=false`;
      if (["", "true", "1", "on", "enable", "enabled"].includes(low)) return `${key}=true`;
      throw new Error(`${key} requires a boolean`);
    } else if (INTEGERS.has(key)) {
      const [min, max] = INTEGERS.get(key);
      if (!/^[0-9]+$/.test(value) || BigInt(value) < min || BigInt(value) > max)
        throw new Error(`${key} requires an integer in [${min}, ${max}]`);
    } else if (key === "--fingerprint-device-memory") {
      if (!/^(?:[0-9]+(?:\.[0-9]*)?|\.[0-9]+)$/.test(value) || !Number.isFinite(Number(value)) ||
          Number(value) <= 0 || Number(value) > 32)
        throw new Error("--fingerprint-device-memory requires a number greater than 0 and at most 32");
    } else if (["--fingerprint-brand-version", "--fingerprint-platform-version"].includes(key)) {
      if (!/^[0-9]+(?:\.[0-9]+){0,3}$/.test(value) || value.split(".").some((n) => BigInt(n) > 0xffffffffn))
        throw new Error(`${key} requires a numeric version with at most four components`);
      if (key === "--fingerprint-brand-version" && (BigInt(value.split(".")[0]) < 1n || BigInt(value.split(".")[0]) > 0x7fffffffn))
        throw new Error(`${key} requires a positive int32 major version`);
    }
    return arg;
  });
  if (fingerprintOff(result))
    result = result.filter((arg) => arg.split("=", 1)[0] !== "--fingerprint-platform");
  validatePreferences(result, final);
  return result;
}

File: sdk/python/chromix/_fingerprint.py
"""Public fingerprint-switch syntax shared by all Python launch paths."""
from __future__ import annotations

import math
import re

OFF_VALUES = frozenset(("off", "false", "0", "disable", "disabled"))
_BRANDS = {"chrome": "Chrome", "google chrome": "Chrome", "edge": "Edge",
           "microsoft edge": "Edge", "opera": "Opera", "vivaldi": "Vivaldi"}
_BOOLEANS = {"--fingerprint-noise", "--fingerprint-sapi-voices",
             "--fingerprint-allow-3p-cookies", "--fingerprint-windows-font-metrics"}
_INTEGERS = {
    "--fingerprint-hardware-concurrency": (1, 128),
    "--fingerprint-screen-width": (1, 32768),
    "--fingerprint-screen-height": (1, 32768),
    "--fingerprint-taskbar-height": (0, 32768),
    "--fingerprint-storage-quota": (0, ((1 << 63) - 1) // (1024 * 1024)),
}
_PREFERENCE_BOOLEANS = {"reduced-motion", "reduced-transparency", "inverted-colors"}
_PREFERENCE_INTEGERS = {"max-touch-points": (0, 16), "timer-resolution": (0, 1000),
                        "audio-seed": (1, (1 << 64) - 1)}
_PREFERENCE_ENUMS = {
    "gpu-backend": {"native", "compatibility"},
    "font-policy": {"native", "restricted"},
    "audio-render": {"native", "isolated"},
    "color-scheme": {"dark", "light"},
    "preferred-contrast": {"more", "less", "no-preference"},
    "forced-colors": {"active", "none", "true", "false", "1", "0"},
    "pointer": {"fine", "coarse", "none"},
    "hover": {"hover", "none"},
    "hdr": {"native"},
    "keyboard-layout": {"native", "us", "en-US"},
}
_CODECS = {"codec-h264", "codec-vp8", "codec-vp9", "codec-av1", "codec-hevc"}


def _valid_codec(value):
    if value in ("", "native", "disabled"):
        return True
    tokens = value.split(",")
    return "supported" in tokens and set(tokens) <= {"supported", "smooth", "power-efficient"}


def _preference(key):
    for prefix in ("--fingerprint-", "--uxr-"):
        if key.startswith(prefix):
            return key[len(prefix):]
    return ""


def _validate_preferences(args, *, final=False):
    # Match Chromium's last-switch semantics and native-alias precedence.
    switches = dict(arg.partition("=")[::2] for arg in args)
    def value(name):
        return switches.get("--uxr-" + name, switches.get("--fingerprint-" + name))
    pointer, hover, touch = value("pointer"), value("hover"), value("max-touch-points")
    touch = int(touch) if touch is not None else None
    if ((pointer == "none" and touch is not None and touch > 0)
            or (pointer == "coarse" and touch == 0)
            or (pointer == "none" and hover == "hover")):
        raise ValueError("inconsistent pointer, hover and max-touch-points")
    keyboard = value("keyboard-layout")
    if keyboard not in (None, "native") and switches.get("--uxr-synthetic-device-tests") != "true":
        raise ValueError("keyboard-layout overrides require synthetic device tests; use native for real input")
    if value("font-policy") == "restricted":
        families = value("font-whitelist") or ""
        entries = families.split(",")
        if (not families or len(families.encode("utf-8")) > 4096 or len(entries) > 256
                or any(not entry.strip(" \t") for entry in entries)
                or re.search(r"[\x00-\x08\x0a-\x1f\x7f]", families)):
            raise ValueError("restricted font-policy requires 1 to 256 comma-separated installed font families")
    if final and value("audio-render") == "isolated" and not fingerprint_off(args):
        seed = value("audio-seed") or switches.get("--uxr-fingerprint-seed")
        if seed is None:
            if "--fingerprint" not in switches:
                raise ValueError("isolated audio requires --fingerprint or a nonzero audio seed")
        elif not re.fullmatch(r"[0-9]+", seed) or not 1 <= int(seed) < (1 << 64):
            raise ValueError("isolated audio requires a nonzero uint64 seed")


def fingerprint_off(args):
    modes = [a.partition("=")[2].lower() for a in args if a.partition("=")[0] == "--fingerprint"]
    return bool(modes and modes[-1] in OFF_VALUES)


def normalize_fingerprint_args(args, *, final=False):
    result = []
    for arg in args or []:
        if not isinstance(arg, str) or "\0" in arg:
            raise ValueError("browser arguments must be strings without NUL")
        key, equal, value = arg.partition("=")
        preference = _preference(key)
        if preference in _PREFERENCE_BOOLEANS:
            low = value.lower()
            if low in OFF_VALUES:
                arg = key + "=false"
            elif low in ("", "true", "1", "on", "enable", "enabled"):
                arg = key + "=true"
            else:
                raise ValueError(f"{key} requires a boolean")
        elif preference in _PREFERENCE_INTEGERS:
            minimum, maximum = _PREFERENCE_INTEGERS[preference]
            if not re.fullmatch(r"[0-9]+", value) or not minimum <= int(value) <= maximum:
                raise ValueError(f"{key} requires an integer in [{minimum}, {maximum}]")
        elif preference in _PREFERENCE_ENUMS:
            if value not in _PREFERENCE_ENUMS[preference]:
                raise ValueError(f"{key} requires one of {', '.join(sorted(_PREFERENCE_ENUMS[preference]))}")
        elif preference in _CODECS:
            if not _valid_codec(value):
                raise ValueError(f"{key} requires native, disabled or supported[,smooth][,power-efficient]")
        elif key == "--fingerprint":
            if value.lower() in OFF_VALUES:
                arg = "--fingerprint=off"
            elif value and (not re.fullmatch(r"[0-9]+", value) or not 1 <= int(value) < (1 << 64)):
                raise ValueError("--fingerprint requires a uint64 seed or off/false/0/disable/disabled")
        elif key == "--fingerprint-brand":
            brand = _BRANDS.get(value.lower())
            if brand is None:
                raise ValueError("--fingerprint-brand must be Chrome, Edge, Opera or Vivaldi")
            arg = key + "=" + brand
        elif key in _BOOLEANS:
            low = value.lower()
            if low in OFF_VALUES:
                arg = key + "=false"
            elif low in ("", "true", "1", "on", "enable", "enabled"):
                arg = key + "=true"
            else:
                raise ValueError(f"{key} requires a boolean")
        elif key in _INTEGERS:
            minimum, maximum = _INTEGERS[key]
            if not re.fullmatch(r"[0-9]+", value) or not minimum <= int(value) <= maximum:
                raise ValueError(f"{key} requires an integer in [{minimum}, {maximum}]")
        elif key == "--fingerprint-device-memory":
            try:
                valid = bool(re.fullmatch(r"(?:[0-9]+(?:\.[0-9]*)?|\.[0-9]+)", value))
                valid = valid and math.isfinite(float(value)) and 0 < float(value) <= 32
            except ValueError:
                valid = False
            if not valid:
                raise ValueError("--fingerprint-device-memory requires a number greater than 0 and at most 32")
        elif key in ("--fingerprint-brand-version", "--fingerprint-platform-version"):
            if not re.fullmatch(r"[0-9]+(?:\.[0-9]+){0,3}", value) or any(int(n) > 0xffffffff for n in value.split(".")):
                raise ValueError(f"{key} requires a numeric version with at most four components")
            if key == "--fingerprint-brand-version" and not 1 <= int(value.split(".")[0]) <= 0x7fffffff:
                raise ValueError(f"{key} requires a positive int32 major version")
        result.append(arg)
    if fingerprint_off(result):
        result = [arg for arg in result if arg.partition("=")[0] != "--fingerprint-platform"]
    _validate_preferences(result, final=final)
    return result
